haystack/pipelines/config.py

```diff
@@ -0,0 +1,119 @@
+            raise KeyError(f"Cannot find any pipeline with name '{pipeline_name}' declared in the YAML file.")
+        pipeline_definition = pipelines_in_definitions[0]
+
+    return pipeline_definition
+
+
+def get_component_definitions(pipeline_config: Dict[str, Any], overwrite_with_env_variables: bool) -> Dict[str, Any]:
+    """
+    Returns the definitions of all components from a given pipeline config.
+
+    :param pipeline_config: Dict Pipeline config parsed as a dictionary.
+    :param overwrite_with_env_variables: Overwrite the YAML configuration with environment variables. For example,
+                                            to change index name param for an ElasticsearchDocumentStore, an env
+                                            variable 'MYDOCSTORE_PARAMS_INDEX=documents-2021' can be set. Note that an
+                                            `_` sign must be used to specify nested hierarchical properties.
+    """
+    component_definitions = {}  # definitions of each component from the YAML.
+    raw_component_definitions = copy.deepcopy(pipeline_config["components"])
+    for component_definition in raw_component_definitions:
+        if overwrite_with_env_variables:
+            _overwrite_with_env_variables(component_definition)
+        name = component_definition.pop("name")
+        component_definitions[name] = component_definition
+
+    return component_definitions
+
+
+def read_pipeline_config_from_yaml(path: Path):
+    with open(path, "r", encoding="utf-8") as stream:
+        return yaml.safe_load(stream)
+
+
+def validate_config(pipeline_config: Dict[str, Any]):
+    for component in pipeline_config["components"]:
+        _validate_user_input(component["name"])
+        _validate_user_input(component["type"])
+        for k, v in component.get("params", {}).items():
+            _validate_user_input(k)
+            _validate_user_input(v)
+    for pipeline in pipeline_config["pipelines"]:
+        _validate_user_input(pipeline["name"])
+        _validate_user_input(pipeline["type"])
+        for node in pipeline["nodes"]:
+            _validate_user_input(node["name"])
+            for input in node["inputs"]:
+                _validate_user_input(input)
+
+
+def build_component_dependency_graph(
+    pipeline_definition: Dict[str, Any], component_definitions: Dict[str, Any]
+) -> DiGraph:
+    graph = DiGraph()
+    for node in pipeline_definition["nodes"]:
+        node_name = node["name"]
+        graph.add_node(node_name)
+        for input in node["inputs"]:
+            if input in component_definitions:
+                graph.add_edge(input, node_name)
+    for component_name, component_definition in component_definitions.items():
+        params = component_definition.get("params", {})
+        referenced_components: List[str] = list()
+        for param_value in params.values():
+            if param_value in component_definitions:
+                referenced_components.append(param_value)
```

Comment:
Problem here is that it will be a bunch of work to validate the signature properly as we are only dealing with strings here:
- resolve type from type name (easy using BaseComponent.subclasses)
- inspect the signature and get the param type
- validate param type is subclass of BaseComponent

This will increase the complexity of the code noticeable. Additionally there is the problem with the last step. We would need to make the assumption that the type annotations in __init__ are:
- correct
- and unambiguous (e.g. not optional or unions or at least we would have to implement these special cases too increasing complexity further)

Furthermore custom types need to be imported before (they need to be registered in BaseComponent, that happens automatically by importing), but do we always want to load them or can we even always load them?

Given that we do not gain too much here, as the example above looks more like an error the user should be notified about rather than building that pipeline and given that such a config would have failed during loading already , I wouldn't introduce it here.
It could however be one of an additional/optional validation steps. E.g. build dependency graph in `validate_config` and check the types of the params against the dependent components.

What do you think?

Reply:
Good points. I agree it would be a complex check and that indeed in most of the cases I can imagine the issue is rather simple to locate. We can probably set it aside for now.

However:
- We definitely need custom components to be loaded by the time we are loading the YAML, so that's not an issue IMHO
- Types should definitely be correct and unambiguous anyway, but I understand it would be quite some time to fix it in all nodes that don't do that. Still it would be a nice push towards improvement :grin: 

I'd say, please add a comment in that area referencing this discussion, so when the time comes we will have a starting point to approach this problem.